app/src/test/java/org/mozilla/fenix/utils/BrowsersCacheTest.kt

```diff
@@ -0,0 +1,146 @@
+import android.content.pm.ActivityInfo
+import android.content.pm.PackageInfo
+import android.content.pm.ResolveInfo
+import android.net.Uri
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import mozilla.components.support.test.robolectric.testContext
+import mozilla.components.support.utils.Browsers
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertNull
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mozilla.fenix.TestApplication
+import org.robolectric.Shadows.shadowOf
+import org.robolectric.annotation.Config
+
+@RunWith(AndroidJUnit4::class)
+@Config(application = TestApplication::class)
+class BrowsersCacheTest {
+
+    // NB: There is always one more browser than pretendBrowsersAreInstalled installs because
+    // the application we are testing is recognized as a browser itself!
+
+    @Test
+    fun testInstallNewBrowsersWithoutReset() {
+        pretendBrowsersAreInstalled(
+            browsers = listOf(
+                Browsers.KnownBrowser.FIREFOX_NIGHTLY.packageName,
+                Browsers.KnownBrowser.REFERENCE_BROWSER.packageName
+            )
+        )
+
+        var browsers = BrowsersCache.all(testContext)
+        assertEquals(browsers.installedBrowsers.size, 3)
+
+        pretendBrowsersAreInstalled(
+            browsers = listOf(
+                Browsers.KnownBrowser.FIREFOX_NIGHTLY.packageName,
+                Browsers.KnownBrowser.FIREFOX.packageName,
+                Browsers.KnownBrowser.CHROME.packageName,
+                Browsers.KnownBrowser.SAMSUNG_INTERNET.packageName,
+                Browsers.KnownBrowser.DUCKDUCKGO.packageName,
+                Browsers.KnownBrowser.REFERENCE_BROWSER.packageName
+            )
+        )
+        browsers = BrowsersCache.all(testContext)
+        assertEquals(browsers.installedBrowsers.size, 3)
+    }
+
+    @Test
+    fun testInstallNewBrowsersWithReset() {
+        pretendBrowsersAreInstalled(
+            browsers = listOf(
+                Browsers.KnownBrowser.FIREFOX_NIGHTLY.packageName,
+                Browsers.KnownBrowser.REFERENCE_BROWSER.packageName
+            )
+        )
+
+        var browsers = BrowsersCache.all(testContext)
+        assertEquals(browsers.installedBrowsers.size, 3)
+
+        pretendBrowsersAreInstalled(
+            browsers = listOf(
+                Browsers.KnownBrowser.FIREFOX_NIGHTLY.packageName,
+                Browsers.KnownBrowser.FIREFOX.packageName,
+                Browsers.KnownBrowser.CHROME.packageName,
+                Browsers.KnownBrowser.SAMSUNG_INTERNET.packageName,
+                Browsers.KnownBrowser.DUCKDUCKGO.packageName,
+                Browsers.KnownBrowser.REFERENCE_BROWSER.packageName
+            )
+        )
+
+        BrowsersCache.resetAll()
+
+        browsers = BrowsersCache.all(testContext)
+        assertEquals(browsers.installedBrowsers.size, 7)
+    }
+
+    @Test
+    fun testResetAll() {
```

Comment:
nit: not strictly necessary but I like to name my tests with the [GIVEN WHEN THEN pattern](https://martinfowler.com/bliki/GivenWhenThen.html): `GIVEN <some test set up> WHEN <event happens> THEN <expected result>`.

It's immediately clear what the test should be testing, which is useful when quickly identifying which tests already exist or when finding the problem when the test fails. It also encourages tests to focus on one bit of functionality. For example, this one might be, `GIVEN the internal cache has a value WHEN resetAll is called THEN the internal cache should be null`.

Reply:
switching to descriptive function names!